staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go

```diff
@@ -0,0 +1,206 @@
+	if cn && fl && hj {
+		return &fancyFailedAuthnResponseWriterDelegator{delegate}
+	}
+	return delegate
+}
+
+var _ http.ResponseWriter = &auditFailedAuthnResponseWriter{}
+
+// auditFailedAuthnResponseWriter intercepts WriteHeader, sets it in the event.
+type auditFailedAuthnResponseWriter struct {
+	http.ResponseWriter
+	event                *auditinternal.Event
+	once                 sync.Once
+	username             string
+	request              *http.Request
+	requestContextMapper request.RequestContextMapper
+	sink                 audit.Sink
+	policy               policy.Checker
+}
+
+func (a *auditFailedAuthnResponseWriter) setHttpHeader() {
+	if a.event != nil {
+		a.ResponseWriter.Header().Set(auditinternal.HeaderAuditID, string(a.event.AuditID))
+	}
+}
+
+func (a *auditFailedAuthnResponseWriter) Write(bs []byte) (int, error) {
+	// the Go library calls WriteHeader internally if no code was written yet. But this will go unnoticed for us
+	a.processCode(http.StatusOK)
+	a.setHttpHeader()
+	return a.ResponseWriter.Write(bs)
+}
+
+func (a *auditFailedAuthnResponseWriter) WriteHeader(code int) {
+	a.processCode(code)
+	a.setHttpHeader()
+	a.ResponseWriter.WriteHeader(code)
+}
+
+func (a *auditFailedAuthnResponseWriter) processCode(code int) {
+	a.once.Do(func() {
+		ctx, ok := a.requestContextMapper.Get(a.request)
+		if !ok {
+			responsewriters.InternalError(a.ResponseWriter, a.request, errors.New("no context found for request"))
+			return
+		}
+
+		// if there already exists an audit event in the context we don't need to do anything
+		if ae := request.AuditEventFrom(ctx); ae != nil {
+			return
+		}
+
+		// otherwise, we need to create the event and log the authn error
+		ev, err := createAuditEvent(ctx, a.policy, a.requestContextMapper, a.request, a.ResponseWriter)
+		if err != nil {
+			return
+		}
+		if ev == nil {
+			return
+		}
+
+		// since user is not set at this point, we need to read it manually
+		ev.User.Username = a.username
+		ctx = request.WithAuditEvent(ctx, ev)
+		if err := a.requestContextMapper.Update(a.request, ctx); err != nil {
+			utilruntime.HandleError(fmt.Errorf("failed to attach audit event to the context: %v", err))
+			responsewriters.InternalError(a.ResponseWriter, a.request, errors.New("failed to update context"))
+			return
+		}
+
+		ev.ResponseStatus = &metav1.Status{}
+		ev.ResponseStatus.Code = int32(code)
+		ev.Stage = auditinternal.StageResponseStarted
+		processAuditEvent(a.sink, ev)
+		a.event = ev
+	})
+}
+
+// getUsername returns username or information on the authn method being used.
+func getUsername(req *http.Request) string {
+	auth := strings.TrimSpace(req.Header.Get("Authorization"))
+
+	// check basic auth
+	const basicScheme string = "Basic "
+	if strings.HasPrefix(auth, basicScheme) {
+		const basic = "<basic>"
+		str, err := base64.StdEncoding.DecodeString(auth[len(basicScheme):])
+		if err != nil {
+			return basic
+		}
+		cred := strings.SplitN(string(str), ":", 2)
```

Comment:
This seems to give an unauthenticated attacker complete control over the username field? Potential XSS vector if downstream consumers are displaying the contents?

Reply:
This piece of code is taken directly from basic auth, so the same problem applies there. If I'm able to address other comment of yours, this should not be a problem anymore.

Reply:
/cc @liggitt

Reply:
Any user can fake a user-name and this name is extracted here to be added to the audit event on authn failure. This by itself is expected and intended. But I agree that allowing any kind of binary data might open up an attack vector on buggy consumers. What is best practice here?

Reply:
same vector applies to resource name parsed from the url, right?

Reply:
For resource names at least the charset is restricted.

Reply:
> For resource names at least the charset is restricted.
not by request parsing, the restrictions are per resource